active/0002-packager.md

```diff
@@ -0,0 +1,257 @@
+
+# Motivation
+
+The current app assembling strategy is naive as it merely merges and concats inputs trees in a nearly undocumented manner. Although this simplicity has served us well, it does limit flexibility of the build process. We believe we can further factor the build process in a way the improves the current À la carte strategy, while providing additional flexibitilty when desired.
+
+# Scope
+
+- Break the build up into 3 distinct phases 
+  - build
+  - satisfy
+  - package
+- Traverse a dependency graph to gain selective builds
+- Lay the infrastructure for the build-side of engines
+- deprecate the hard dependency on Bower in favor NPM
+- deprecate app.import API
+
+# Definitions
+
+- __Builder__ - Responsible for generating an array of broccoli trees that contain the app and all the addons.
+- __Pre-Packager__ - Responsible for resolving and reducing the graph to a tree containing only reachable modules.
+- __Packager__ - Responsible for taking the resolved tree and applying – default or user provided – concatenation strategies to the tree.
+- __Entry__ - A entry node into the graph in which you traverse down from. This can be an app or engine.
+
+# Detailed Design
+
+## Builder
+The Builder is a re-thinking of the existing `EmberApp` constructor function that is used in the Brocfile today.  Unlike the existing `EmberApp` constructor, the `Builder` constructor only does the following:
+
+- Discover the addons
+- Runs the addons hooks that are pre-resolved tree
+- Merges an addon's app directory with the consuming app
+- Transpiles ES2015-modules code to AMD
+
+As part of the ES2015-modules transpilation we drop a `dep-graph.json` at the root of each tree. This is used by the Pre-Packager to resolve the graph. The result of this phase is an array of Broccoli trees containing all of the built assets in the app and addons.
+
+At this point nothing is really different then what happens today besides dropping the `dep-graph.json` and only running the addon hooks that need to happen at the beginning of a build.
+
+## Pre-Packager
+The Pre-Packager is a graph resolver algorithm for javascript dependencies. Another way of think about the Pre-Packager is a complex [Broccoli-Filter](https://github.com/broccolijs/broccoli-filter) and [Broccoli-Merge-Trees](https://github.com/broccolijs/broccoli-merge-trees).
+
+The Pre-Packager can resolve the following types out of the box:
+
+- `ember-addons`
+- Legacy modules from the node_modules directory (CJS)
+- ES2015 modules from the node_modules directory
+
+While the Pre-packager comes with a sane set of defaults, custom dependency resolvers can be written to allow resolution of modules outside of node_modules.
+
+### Dependency Resolvers
+
+There are 2 types of dependency resolvers in the pre-packager.
+
+1. Resolvers that start traversal from a static graph
+2. Resolvers that must wait till the static graph has materialized
```

Comment:
this section would likely be clearer if we had a name for the output of 1. as that is what 2. operators on.

Maybe a name isn't possible, but a definition is possible:

The result of 1. is an entirely reachable graph of app, engine and addon code. This graph includes all engine/app branches, but only reachable add-on branches. If this graph has missing entries – is unsatistied – we proceed to attempt to satisfy the missing dependencies from foreign sources, such as CJS or npm provided ES2015.